### PR TITLE
feat: Qdrant datetime filtering support

### PR DESCRIPTION
**Pull request by Anush008**
Qdrant v1.8 introduced support for [datetime filtering](https://qdrant.tech/documentation/concepts/filtering/#datetime-range).

This PR updates the Qdrant integration to support the same in Haystack.

### CODE REVIEW

**Comment by Anush008** [comment]
Blocked by https://github.com/qdrant/qdrant-client/issues/532.

**Comment by Anush008** [comment]
Hi @anakin87. Could you please rerun the CI?

`qdrant-client==1.8.1` should now fix this.

File: integrations/qdrant/src/haystack_integrations/document_stores/qdrant/filters.py [hunk 1]
```diff
@@ -141,34 +133,47 @@ def _build_nin_condition(self, key: str, value: List[models.ValueVariants]) -> m
             must_not=[
                 (
                     models.FieldCondition(key=key, match=models.MatchText(text=item))
-                    if isinstance(item, str) and " " not in item
+                    if isinstance(item, str) and " " in item
```
Comment (anakin87) [inline integrations/qdrant/src/haystack_integrations/document_stores/qdrant/filters.py:136]: can you explain this change?

File: integrations/qdrant/src/haystack_integrations/document_stores/qdrant/filters.py [hunk 2]
```diff
@@ -141,34 +133,47 @@ def _build_nin_condition(self, key: str, value: List[models.ValueVariants]) -> m
...
 
-    def _build_lt_condition(self, key: str, value: float) -> models.Condition:
+    def _build_lt_condition(self, key: str, value: Union[str, float, int]) -> models.Condition:
+        if isinstance(value, str) and is_datetime_string(value):
+            return models.FieldCondition(key=key, range=models.DatetimeRange(lt=value))
+
         if not isinstance(value, (int, float)):
-            msg = f"Value {value} is not an int or float"
+            msg = f"Value {value} is not an int or float or datetime string"
             raise FilterError(msg)
```
Comment (anakin87) [inline integrations/qdrant/src/haystack_integrations/document_stores/qdrant/filters.py:149]: I would do something like this:
```python
    msg = f"Value {value} is not an int or float or datetime string"
    if isinstance(value, str):
        if is_datetime_string(value):
            return models.FieldCondition(key=key, range=models.DatetimeRange(lt=value))
       raise FilterError(msg)

        if not isinstance(value, (int, float)):
          raise FilterError(msg)
```

It is important that if the string is not recognized as a datetime object, an error is raised.

**Review by anakin87 — COMMENTED** [review]
I'm happy that the tests pass.
Thanks for this contribution!

I left two comments.

File: integrations/qdrant/src/haystack_integrations/document_stores/qdrant/filters.py (hunk 1 above)
Comment (Anush008) [reply integrations/qdrant/src/haystack_integrations/document_stores/qdrant/filters.py:136]: This fixes the previous faulty implementation.
If the string has a whitespace we do a [full-text match](https://qdrant.tech/documentation/concepts/filtering/#full-text-match) else a [keyword/number match](https://qdrant.tech/documentation/concepts/filtering/#match).

File: integrations/qdrant/src/haystack_integrations/document_stores/qdrant/filters.py (hunk 2 above)
Comment (Anush008) [reply integrations/qdrant/src/haystack_integrations/document_stores/qdrant/filters.py]: Would https://github.com/deepset-ai/haystack-core-integrations/pull/570/commits/dd821d05269183a899b3b97cd88828790f9931e5 do?

**Review by anakin87 — APPROVED** [review]
LGTM!

**Comment by anakin87** [comment]
Released: https://pypi.org/project/qdrant-haystack/3.2.0/

**Pull request merged**